Autopilot/AttitudeManager/InputCapture.c

```diff
@@ -1,298 +1,318 @@
-        IEC2bits.IC4IE = 1; // Enable IC4 interrupt
+void __attribute__((__interrupt__, no_auto_psv)) _IC6Interrupt(void)
+{
+    if (PORTDbits.RD13 == 1) {
+        start_time[5] = IC6BUF;
+        new_data_available[5] = 0;
+    } else {
+        end_time[5] = IC6BUF;
+        new_data_available[5] = 1;
     }
-    if (initIC & 0b10000){
-        IC5CONbits.ICM = 0b00; // Disable Input Capture 5 module
-        IC5CONbits.ICTMR = 1; // Select Timer2 as the IC1 Time base
-        IC5CONbits.ICI = 0b11; // Interrupt on every capture event
-        IC5CONbits.ICM = 0b001; // Generate capture event on every Rising edge
-
-        // Enable Capture Interrupt And Timer2
-        IPC9bits.IC5IP = 7; // Setup IC5 interrupt priority level - Highest
-        IFS2bits.IC5IF = 0; // Clear IC5 Interrupt Status Flag
-        IEC2bits.IC5IE = 1; // Enable IC5 interrupt
+    
+    while (IC6CONbits.ICBNE) {
+        IC6BUF;
     }
-    if (initIC & 0b100000){
-        IC6CONbits.ICM = 0b00; // Disable Input Capture 6 module
-        IC6CONbits.ICTMR = 1; // Select Timer2 as the IC1 Time base
-        IC6CONbits.ICI = 0b11; // Interrupt on every capture event
-        IC6CONbits.ICM = 0b001; // Generate capture event on every Rising edge
+    IFS2bits.IC6IF = 0;
+}
 
-        // Enable Capture Interrupt And Timer2
-        IPC10bits.IC6IP = 7; // Setup IC6 interrupt priority level - Highest
-        IFS2bits.IC6IF = 0; // Clear IC6 Interrupt Status Flag
-        IEC2bits.IC6IE = 1; // Enable IC6 interrupt
-    }
-    if (initIC & 0b1000000){
-        IC7CONbits.ICM = 0b00; // Disable Input Capture 7 module
-        IC7CONbits.ICTMR = 1; // Select Timer2 as the IC1 Time base
-        IC7CONbits.ICI = 0b11; // Interrupt on every capture event
-        IC7CONbits.ICM = 0b001; // Generate capture event on every Rising edge
+//Input Capture #7 Interrupt Function
 
-        // Enable Capture Interrupt And Timer2
-        IPC5bits.IC7IP = 7; // Setup IC7 interrupt priority level - Highest
-        IFS1bits.IC7IF = 0; // Clear IC7 Interrupt Status Flag
-        IEC1bits.IC7IE = 1; // Enable IC7 interrupt
+void __attribute__((__interrupt__, no_auto_psv)) _IC7Interrupt(void)
+{
+    if (PORTDbits.RD14 == 1) {
+        start_time[6] = IC7BUF;
+        new_data_available[6] = 0;
+    } else {
+        end_time[6] = IC7BUF;
+        new_data_available[6] = 1;
     }
-    if (initIC & 0b10000000){
-        IC8CONbits.ICM = 0b00; // Disable Input Capture 8 module
-        IC8CONbits.ICTMR = 1; // Select Timer2 as the IC1 Time base
-        IC8CONbits.ICI = 0b11; // Interrupt on every capture event
-        IC8CONbits.ICM = 0b001; // Generate capture event on every Rising edge
 
-        // Enable Capture Interrupt And Timer2
-        IPC5bits.IC8IP = 7; // Setup IC8 interrupt priority level - Highest
-        IFS1bits.IC8IF = 0; // Clear IC8 Interrupt Status Flag
-        IEC1bits.IC8IE = 1; // Enable IC8 interrupt
+    while (IC7CONbits.ICBNE) {
```

Comment:
So the pic has a FIFO buffer of size 4 that stores a the timer values.  When you read from IC1BUF you're actually retrieving one of the values from the buffer, and removing it (so shrinking it by one). The ICBNE register is 1 if the buffer is not empty (if there are values to read from it). In our case, since we trigger the interrupt on every edge, the size of the buffer should only ever be 1. However I noticed some strange behaviour when I unplugged a PWM cable from the input and plugged it back in. It seemed to have been an issue with the buffer still containing values, thus it was reading the wrong things. Im still not 100% sure, but I know that with this code I didn't encounter those issues.